### ISSUE TITLE
jax.lax.custom_linear_solve / jax.lax.custom_root doesn't respect jax.disable_jit

### ISSUE THREAD

**Issue by YouJiacheng** [google/jax#10022]
After #9938 
The document says:
> For debugging it is useful to have a mechanism that disables [jit()](https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html#jax.jit) everywhere in a dynamic context. Note that this not only disables explicit uses of jit by the user, but will also remove any implicit JIT compilation used by the JAX library: this includes implicit JIT computation of body and cond functions passed to higher-level primitives like scan() and while_loop(), JIT used in implementations of [jax.numpy](https://jax.readthedocs.io/en/latest/jax.numpy.html#module-jax.numpy) functions, and any other case where jit is used within an API’s implementation.

However, when I worked for https://github.com/google/jax/issues/9714#issuecomment-1077319705, I found that there are jaxprs generated inside `with jax.disable_jit():`. 
I checked the source of `jax.lax.custom_linear_solve`, and found it doesn't check `config.jax_disable_jit`.
`jax.lax.custom_root` has the same problem too.

**Comment by mattjj** [comment]
Thanks for raising this. Maybe the docs should say it disables jit in many (but not all) places. Then we can treat it as an enhancement to support this feature in more cases.

@shoyer does it make sense to make `custom_linear_solve` / `custom_root` sensitive to `disable_jit`?

@YouJiacheng jaxprs may still be generated even with `disable_jit` supported everywhere because we use them both for `jit`-style staging of computations as well as reverse-mode AD.

**Comment by YouJiacheng** [comment]
Thanks for reply.
I think at least we can print a INFO or WARN when jit isn't disabled. Because `jax.disable_jit` is intended for debugging, adding some outputs seems reasonable.

**Comment by shoyer** [comment]
custom_linear_solve/custom_root trace to a JAXpr using the equivalent of jax.closure_convert. It isn't really tractable to evaluate JAXprs without JIT, because loops become extremely slow (way slower than pure Python).

It would be nice to rewrite these functions to avoid using closure conversion automatically, which would make them more flexible and make using disable JIT feasible. This might require tweaking the APIs slightly -- I haven't really thought about it yet.

**Comment by patrick-kidger** [comment]
If it's a helpful point of reference. Diffrax uses its own custom-jvp-through-root-finding code for nonlinear solves, as I didn't want the magic (and previously, bugs) of `closure_convert`-type behaviour. So this has an API that may serve as inspiration.

See [here](https://github.com/patrick-kidger/diffrax/blob/1898345b7b0725df0d8776bf8f02657a9182568f/diffrax/nonlinear_solver/base.py#L105) for the explicit commentary; more broadly the rest of the `nonlinear_solver` folder.

**Comment by shoyer** [comment]
@patrick-kidger interesting, thanks for sharing! One important difference between your solver and `lax.custom_root` is that you calculate a full Jacobian and use `jnp.linalg.solve`, where `lax.custom_root` allows for a custom (possibly matrix-free) linear solver.

**Comment by patrick-kidger** [comment]
Absolutely. I'm not claiming that it's a perfectly equivalent API, just that it might make a useful point of reference as another take on this, in particular without `closure_convert`.
(FWIW I'll probably add support for other linear solvers at some point down the line, if/when that becomes necessary.)

**Comment by nicholasjng** [comment]
Just wanted to chime in that I'm happy to take a look at these, I've had some fun with adding auxiliary arguments to `custom_{root,linear_solve}` in the past and would be happy to continue.

**Comment by shoyer** [comment]
@nicholasjng Contributions here would be very welcome! I'm not 100% sure it will be possible `custom_linear_solve` because we currently need implement `custom_linear_solve` with a JAX primitive (until https://github.com/google/jax/issues/9129 implementing `custo_transpose` is finished, CC @froystig). It might be worth holding off any API changes until we can switch both `custom_linear_solve` and `custom_root` at once, but in any case starting to investigate this would certainly be valuable.

**Comment by aadityacs** [comment]
I am trying to use [PyAMG](https://pyamg.readthedocs.io/en/latest/generated/pyamg.krylov.html) for solving my PDEs. I thought I could wrap the solver inside `jax.lax.custom_linear_solve` to ensure that I can also perform the adjoint solve. 

However, it seems that JAX throws a `TracerArrayConversionError` as it tries to differentiate via the solver. However, documentation mentions that the solver need not be differentiable. 

I am not sure if I am doing something wrong. I was hoping someone could kindly point me in the right direction. Thanks

A standalone simple implementation can be found [here](https://colab.research.google.com/drive/1aKCudyjzBXVN7_ft0rh4IrYacGBIm5Ok?usp=sharing)

**Comment by shoyer** [comment]
You will need to use an [external callback](https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html) to wrap libraries for NumPy arrays like PyAMG with `custom_linear_solve`.
